geniza/corpus/management/commands/add_fragment_urls.py

```diff
@@ -0,0 +1,144 @@
+        self.stdout.write(f"Fragments not found: {self.stats['not_found']}")
+        self.stdout.write(f"Fragments skipped: {self.stats['skipped']}")
+
+    def view_to_iiif_url(self, url):
+        """Generate IIIF Manifest URL based on view url, if it can
+        be determined automatically"""
+
+        # cambridge iiif manifest links use the same id as view links
+        # NOTE: should exclude search link like this one:
+        # https://cudl.lib.cam.ac.uk/search?fileID=&keyword=T-s%2013J33.12&page=1&x=0&y=
+        if "cudl.lib.cam.ac.uk/view/" in url:
+            iiif_link = url.replace("/view/", "/iiif/")
+            # view links end with /1 or /2 but iiif link does not include it
+            iiif_link = re.sub(r"/\d$", "", iiif_link)
+            return iiif_link
+
+        return ""
+
+    def add_fragment_urls(self, row):
+        try:
+            fragment = Fragment.objects.get(shelfmark=row["shelfmark"])
+        except Fragment.DoesNotExist:
+            self.stats["not_found"] += 1
+            return
+
+        url = row.get("url")
+        iiif_url = row.get("iiif_url") or self.view_to_iiif_url(row["url"])
+        save_needed = False
+        log_message = []
+
+        # if there is a view url, add or optionally update it
+        if url:
+            if not fragment.url:
+                fragment.url = url
+                self.stats["url_added"] += 1
+                save_needed = True
+                log_message.append("added URL")
+            elif fragment.url != url:
+                self.stdout.write(
+                    "Fragment %s url differs: %s %s" % (fragment, fragment.url, url)
+                )
+                if self.overwrite:
+                    fragment.url = url
+                    self.stats["url_updated"] += 1
+                    save_needed = True
+                    log_message.append("updated URL")
+
+        # similar logic for iiif url
+        if iiif_url:
+            if not fragment.iiif_url:
+                fragment.iiif_url = iiif_url
+                self.stats["iiif_added"] += 1
+                save_needed = True
+                log_message.append("added IIIF URL")
+            elif fragment.iiif_url != iiif_url:
+                self.stdout.write(
+                    "Fragment %s IIIF url differs: %s %s"
+                    % (fragment, fragment.iiif_url, iiif_url)
+                )
+                if self.overwrite:
+                    self.stats["iiif_updated"] += 1
+                    fragment.iiif_url = iiif_url
+                    save_needed = True
+                    log_message.append("updated IIIF URL")
+
+        if save_needed:
+            if self.dryrun:
+                self.stdout.write(f"Set {fragment} url to {url} and iiif to {iiif_url}")
+            else:
+                fragment.save()
+                self.log_change(fragment, " and ".join(log_message))
+        else:
+            self.stats["skipped"] += 1
```

Comment:
this is nice; i like avoiding the `.save()` call on things we don't change. probably speeds it up too